feat(table): add fanout partition writer and rolling data writer

badalprasadsingh opened the pull request:
# Partitioned Fanout Writer with Rolling Data File Support (Append Mode)

This PR completes the implementation of partitioned writing with support for rolling data files in append mode. It enables efficient, parallelized ingestion into partitioned tables while maintaining manifest and snapshot correctness.

**Slack Thread Discussion**: [Link](https://apache-iceberg.slack.com/archives/C05J3MJ42BD/p1751002533414969)
**Proposal Document**: [Google Drive](https://drive.google.com/file/d/18CwR9nhwkThs-Q-JZZvisBEaDICvp5Z7/view?usp=drive_link)


### Details

* Introduced parallel processing of `arrow.Record` using a user-defined number of goroutines.
* Each goroutine maintains its own hash table to map partition keys to row indices.
* After partitioning, `compute.Take()` is used to extract per-partition data slices.
* Integrated dedicated rolling writers per partition to manage data file size thresholds and output constraints.


### Tests Performed

* [x] Compatible with all partition transforms
* [x] Handled null values in partition columns
* [x] Validated compatibility with partition spec evolution
* [x] Verified correctness for non-linear transformation cases
* [x] Confirmed schema evolution compatibility
* [x] Partition pruning verified

---

@zeroshade — would appreciate your review when you get a chance!

badalprasadsingh committed:
feat(writer): add fanout partition writer and rolling data writer

badalprasadsingh committed:
fix: fanout writer and rolling data writer

badalprasadsingh committed:
fixed partitioned writer and added tests

badalprasadsingh committed:
fixed flushing logic and error handling

badalprasadsingh committed:
fanout writer: minor fixes

badalprasadsingh committed:
fixed race conditions

badalprasadsingh committed:
fixed for non linear transformations

badalprasadsingh committed:
Merge branch 'main' into feature/fanout-partition-writer

badalprasadsingh committed:
add: test for fanout partition writer

badalprasadsingh committed:
fix: minor issues

badalprasadsingh committed:
add: function signatures

badalprasadsingh committed:
feat(table): add fanout partition writer and rolling data writer

This completes the implementation of partitioned writing with support for rolling data files in append mode. It enables efficient, parallelized ingestion into partitioned tables while maintaining manifest and snapshot correctness.

Details

- Introduced parallel processing of `arrow.Record`s using a user-defined number of goroutines.
- Each goroutine maintains its own hash table to map partition keys to row indices.
- After partitioning, `compute.Take()` is used to extract per-partition data slices.
- Integrated dedicated rolling writers per partition to manage data file size thresholds and output constraints.

Signed-off-by: badalprasadsingh <badal@datazip.io>

kevinjqliu commented:
Thanks for the PR @badalprasadsingh could you address the linter issue in CI?

zeroshade reviewed (changes requested):
Did a first pass and left some comments for start. Will need to do another tomorrow

badalprasadsingh committed:
fix: resolve review comments from PR feedback

Signed-off-by: badalprasadsingh <badal@datazip.io>

badalprasadsingh committed:
fix: resolve review comments from PR feedback

Signed-off-by: badalprasadsingh <badal@datazip.io>

badalprasadsingh commented on schema_conversions.go:
```diff
@@ -30,33 +30,35 @@ func partitionTypeToAvroSchema(t *StructType) (avro.Schema, error) {
 		switch typ := f.Type.(type) {
 		case Int32Type:
-			sc = internal.IntSchema
+			sc = internal.NullableSchema(internal.IntSchema)
```
This change ensures the schema allows nullable data. Without it, if the entire column contains `null` values, it triggers the error: `PartitionData: col: avro: *avro.null`. Wrapping inside `internal.NullableSchema` prevents this issue.

badalprasadsingh replied:
Sure

badalprasadsingh replied:
Yes you're absolutely right!

badalprasadsingh replied:
Yes we can do that.

badalprasadsingh committed:
fix: address review feedback and improve logical type conversion in Avro

Updated Avro serialization logic to correctly handle decimal logical types and applied suggested code refactoring from PR review.

Signed-off-by: badalprasadsingh <badal@datazip.io>

badalprasadsingh committed:
fix: resolve review comments and logical types conversion in avro

Signed-off-by: badalprasadsingh <badal@datazip.io>

badalprasadsingh committed:
fix: minor issues

Signed-off-by: badalprasadsingh <badal@datazip.io>

badalprasadsingh committed:
merge main

Signed-off-by: badalprasadsingh <badal@datazip.io>

badalprasadsingh committed:
fix: logical types conversion in avro

Signed-off-by: badalprasadsingh <badal@datazip.io>

badalprasadsingh commented on table/arrow_utils.go:
```diff
@@ -1307,8 +1309,19 @@ func recordsToDataFiles(ctx context.Context, rootLocation string, meta *Metadata
+		workers := meta.props.GetInt(FanoutWriterWorkersKey, FanoutWriterWorkersDefault)
+		if workers <= 0 {
+			workers = runtime.NumCPU()
+		}
```
Fixed

badalprasadsingh commented on table/partitioned_fanout_writer.go:
```diff
@@ -0,0 +1,309 @@
+		}
+	}
+
+	for i := 0; i < workers; i++ {
```
Fixed

badalprasadsingh commented on table/partitioned_fanout_writer.go:
```diff
@@ -0,0 +1,309 @@
+		return nil
+	})
+
+	return nil
```
Fixed

badalprasadsingh commented on table/partitioned_fanout_writer.go:
```diff
@@ -0,0 +1,309 @@
+			if err := context.Cause(ctx); err != nil {
+				return err
+			}
+			return nil
```
Fixed

badalprasadsingh commented on table/partitioned_fanout_writer.go:
```diff
@@ -0,0 +1,309 @@
+				if err := context.Cause(ctx); err != nil {
+					return err
+				}
+				return nil
```
Fixed

badalprasadsingh commented on table/partitioned_fanout_writer.go:
```diff
@@ -0,0 +1,309 @@
+						return err
+					}
+					return nil
+				default:
```
Fixed

badalprasadsingh commented on table/partitioned_fanout_writer.go:
```diff
@@ -0,0 +1,309 @@
+}
+
+func (p *PartitionedFanoutWriter) yieldDataFiles(ctx context.Context, fanoutWorkers *errgroup.Group, outputDataFilesCh chan iceberg.DataFile) iter.Seq2[iceberg.DataFile, error] {
+	return iter.Seq2[iceberg.DataFile, error](func(yield func(iceberg.DataFile, error) bool) {
```
Fixed

badalprasadsingh commented on table/partitioned_fanout_writer.go:
```diff
@@ -0,0 +1,309 @@
+		}()
+
+		<-done
+		<-bufferDone
```
Fixed

badalprasadsingh commented on table/partitioned_fanout_writer.go:
```diff
@@ -0,0 +1,309 @@
+	partitionFields := p.partitionSpec.PartitionType(p.schema).FieldList
+
+	for row := range record.NumRows() {
+		partitionRec := make(partitionRecord, len(partitionFields))
```
Fixed

badalprasadsingh committed:
update: properties.go

Signed-off-by: badalprasadsingh <badal@datazip.io>

badalprasadsingh committed:
fix: schema_conversions_test.go

Signed-off-by: badalprasadsingh <badal@datazip.io>

zeroshade commented:
Hey @badalprasadsingh I'm having some Internet issues but I should be able to review this tomorrow

badalprasadsingh committed:
fix: staticcheck

Signed-off-by: badalprasadsingh <badal@datazip.io>

badalprasadsingh committed:
fix: padOrTruncateBytes and unexport partition writer types

Signed-off-by: badalprasadsingh <badal@datazip.io>

zeroshade commented on table/rolling_data_writer.go:
```diff
@@ -0,0 +1,199 @@
+	}
+}
+
+func (w *WriterFactory) getOrCreateRollingDataWriter(partition string, partitionValues map[int]any) (*RollingDataWriter, error) {
```
We shouldn't need to buffer all of the records, just create a new writer on the fly as we encounter new partitions. The `iter.Seq2` objects can be tied to channels, mostly the reason why I mention `recordToDataFiles` is the `binPackRecords` function being used. So we dont' necessarily need to call `recordsToDataFiles` directly but rather have `getOrCreateRollingDataWriter` leverage the `binPackRecords` function for better packing of the records to get more efficient writes to the files.

I'll try to put together a flow-chart or something to better explain what I'm talking about. We just need to buffer enough records to perform the binPacking for each partition

badalprasadsingh committed:
fix: getArrowValueAsIcebergLitera func()

Signed-off-by: badalprasadsingh <badal@datazip.io>

badalprasadsingh committed:
fix: filesToDataFiles func()

Signed-off-by: badalprasadsingh <badal@datazip.io>

badalprasadsingh committed:
fix: rolling data writer

Signed-off-by: badalprasadsingh <badal@datazip.io>

badalprasadsingh committed:
fix: manifest_test.go

Signed-off-by: badalprasadsingh <badal@datazip.io>

badalprasadsingh commented on manifest.go:
```diff
@@ -1461,33 +1520,116 @@ func mapToAvroColMap[K comparable, V any](m map[K]V) *[]colMap[K, V] {
+		return v
+	}
+
+	byteSize := 5
```
Fixed

badalprasadsingh committed:
fix: partition summaries & rolling data writer

Signed-off-by: badalprasadsingh <badal@datazip.io>

zeroshade commented on manifest.go:
```diff
@@ -37,6 +37,19 @@ import (
 	"github.com/hamba/avro/v2/ocf"
 )
 
+func init() {
+	avro.Register("fixed", [1]byte{})
+	avro.Register("fixed", [2]byte{})
+	avro.Register("fixed", [3]byte{})
+	avro.Register("fixed", [4]byte{})
+	avro.Register("fixed", [5]byte{})
+	avro.Register("fixed", [6]byte{})
+	avro.Register("fixed", [7]byte{})
+	avro.Register("fixed", [8]byte{})
+	avro.Register("fixed", [16]byte{})
```
According to the `hambra` documentation, it should automatically decode `fixed` to `[n]byte` without needing these, do we have a test for this?

badalprasadsingh replied:
Yes, you are right but, when we have a `nullable` fixed schema like:
```json
["null",{"name":"fixed","type":"fixed","size":16}]
```

The hamba/avro library needs to know how to map Go types to the Avro schema types. For fixed schemas, it expects specific byte array types like [16]byte, but the library's type resolution system couldn't automatically handle all the different fixed sizes.

Doing `avro.Register`, says that when it sees a [N]byte type, treat it as a 'fixed' schema type for `union` resolution purposes.

One possible fix for it is for fixed type use:
```go
internal.NullableSchema(internal.BinarySchema)
```
instead of 
```go
fixedSchema := internal.Must(avro.NewFixedSchema("fixed", "", typ.len, nil))
sc = internal.NullableSchema(fixedSchema)
```

zeroshade replied:
that seems like a good solution to me, go for it

zeroshade commented on manifest.go:
```diff
@@ -416,30 +429,53 @@ func getFieldIDMap(sc avro.Schema) (map[string]int, map[int]avro.LogicalType) {
+		result[field.Name()] = fid
+		avroTyp := field.Type()
+		if us, ok := avroTyp.(*avro.UnionSchema); ok {
+			for _, t := range us.Types() {
+				avroTyp = t
 			}
```
should we confirm that we're not using null?

badalprasadsingh replied:
Yes we are sure. Done.

zeroshade commented on manifest.go:
```diff
@@ -960,6 +1003,37 @@ func (p *partitionFieldStats[T]) update(value any) (err error) {
 }
 
+func extractBytesFromFixed(fixedBytes interface{}) []byte {
+	switch fb := fixedBytes.(type) {
+	case []interface{}:
```
What situation does this happen with? Where we get `[]interface{}` instead of `[]byte`?

badalprasadsingh replied:
Yes you are right! Fixed it.

zeroshade commented on manifest.go:
```diff
@@ -1461,33 +1598,114 @@ func mapToAvroColMap[K comparable, V any](m map[K]V) *[]colMap[K, V] {
+	}
+
+	if bytes, ok := v.([]byte); ok && fixedSize > 0 {
+		return convertToFixedArray(padOrTruncateBytes(bytes, fixedSize), fixedSize)
```
why do we need to do this? We can't just return the uuid?

badalprasadsingh replied:
Done.

zeroshade commented on manifest.go:
```diff
@@ -1461,33 +1598,114 @@ func mapToAvroColMap[K comparable, V any](m map[K]V) *[]colMap[K, V] {
+	if err != nil {
+		return v
+	}
+	fixedArray := convertToFixedArray(padOrTruncateBytes(bytes, fixedSize), fixedSize)
```
why converting to fixed array instead of just returning the decimal type?

zeroshade replied:
Then that should be happening inside of `DataFileStatistics.PartitionValue` or in `DataFileStatistics.ToDataFile`. That's where we perform the writes. This code path only happens on read, so I don't know why we're doing this here.

zeroshade commented on table/arrow_utils.go:
```diff
@@ -1030,11 +1031,19 @@ func (sc *schemaCompatVisitor) isFieldCompat(lhs iceberg.NestedField) bool {
-		tbl.Render()
-		txt, _ := tbl.Srender()
-		pterm.EnableColor()
+		var lines []string
+		lines = append(lines, "   | Table Field              | Requested Field")
+
+		for i, row := range sc.errorData {
+			if i == 0 {
+				continue
+			}
+			if len(row) >= 3 {
+				lines = append(lines, fmt.Sprintf("%s | %-24s | %s", row[0], row[1], row[2]))
+			}
+		}
```
why the change here?

badalprasadsingh replied:
This change fixes the formatting mismatch for `TestTableWriting` test.

zeroshade reviewed (changes requested):
A few more questions, also there are still some of my earlier comments waiting for responses.

Otherwise this is shaping up really well

badalprasadsingh commented on exprs.go:
```diff
@@ -482,8 +482,33 @@ func (b *boundRef[T]) Equals(other BoundTerm) bool {
-func (b *boundRef[T]) Type() Type          { return b.field.Type }
+
+func unwrapLogicalTypeValue(v any) any {
+	if m, ok := v.(map[string]any); ok {
+		if val, exists := m["long.timestamp-micros"]; exists {
+			if microseconds, ok := val.(int64); ok {
+				return Timestamp(microseconds)
+			}
+		}
+
+		if val, exists := m["int.date"]; exists {
+			if days, ok := val.(int32); ok {
+				return days
+			}
+		}
+
+		if val, exists := m["long.time-micros"]; exists {
+			if microseconds, ok := val.(int64); ok {
+				return Time(microseconds)
+			}
+		}
+	}
+
+	return v
```
well, when we have a union schema like 
```json
["null", {"type": "int", "logicalType": "date"}]
```
the Avro library needs to determine which type in the union to use when encoding a value. If you just pass a raw `int32` value, the library can't distinguish.

That's the reason why we have:
```go
return map[string]any{"int.date": int32(d)}
```
in `convertDateValue` function. Thus, later on we have had to convert them back to `iceberg.Literal` formats

zeroshade replied:
the conversion should happen inside of `ToDataFile` not here. We should never have the `map` representation here. Specifically, it should be happening in `DataFileStatistics.PartitionValue`, that's where we should convert from the literal to the map representation.

badalprasadsingh commented on table/internal/utils.go:
```diff
@@ -212,8 +214,7 @@ func (d *DataFileStatistics) PartitionValue(field iceberg.PartitionField, sc *ic
 	}
 
 	if !field.Transform.PreservesOrder() {
-		panic(fmt.Errorf("cannot infer partition value from parquet metadata for a non-linear partition field: %s with transform %s",
-			field.Name, field.Transform))
+		return nil
```
From the `partitioned-fanout-writer` side I pass back the partition values for all transformations. So in case it encounters a `non-linear` transformation, it can use the partition value from the writer side for each `datafile`, no need for it to calculate.

However in case of `AddFiles()` func which ultimately calls the `filesToDataFiles` func, we still throw an `error` for non-linear transformations

badalprasadsingh commented on manifest.go:
```diff
@@ -1461,33 +1520,116 @@ func mapToAvroColMap[K comparable, V any](m map[K]V) *[]colMap[K, V] {
+	}
+}
+
+func convertDateValue(v any) any {
+	if t, ok := v.(time.Time); ok {
+		return map[string]any{"int.date": int32(t.Truncate(24*time.Hour).Unix() / int64((time.Hour * 24).Seconds()))}
+	}
+	if d, ok := v.(Date); ok {
+		return map[string]any{"int.date": int32(d)}
+	}
+
+	return v
+}
```
Well, since we are using `nullable` schema, the map pattern like 
```go
map[string]any{"int.date": int32(value)}
```
was created as a way to explicitly tag which union member should be used:

```go
return map[string]any{"int.date": int32(d)}
```
The key "int.date" serves as a type hint to resolve the union ambiguity.

badalprasadsingh commented on table/partitioned_fanout_writer.go:
```diff
@@ -0,0 +1,320 @@
+
+// NewPartitionedFanoutWriter creates a new PartitionedFanoutWriter with the specified
+// partition specification, schema, and record iterator.
+func NewPartitionedFanoutWriter(partitionSpec iceberg.PartitionSpec, schema *iceberg.Schema, itr iter.Seq2[arrow.Record, error]) *PartitionedFanoutWriter {
```
Fixed

badalprasadsingh committed:
fix: avro logical types (nullable)

Signed-off-by: badalprasadsingh <badal@datazip.io>

zeroshade commented on manifest.go:
```diff
@@ -416,30 +429,53 @@ func getFieldIDMap(sc avro.Schema) (map[string]int, map[int]avro.LogicalType) {
-					avroTyp = t
-				}
+		var fid int
+		switch v := field.Prop("field-id").(type) {
```
Go JSON parsing should always return a number as a `float64` i believe. Do we have any tests that hit any scenarios other than `float64` for this?

zeroshade commented on manifest.go:
```diff
@@ -1461,33 +1574,128 @@ func mapToAvroColMap[K comparable, V any](m map[K]V) *[]colMap[K, V] {
+		return map[string]any{"null": nil}
+	}
+
+	if uuidVal, ok := v.(uuid.UUID); ok {
+		uuidLiteral := UUIDLiteral(uuidVal)
+		bytes, _ := uuidLiteral.MarshalBinary()
+		fixedArray := convertToFixedArray(padOrTruncateBytes(bytes, 16), 16)
+
+		return map[string]any{"uuid": fixedArray}
```
`uuid.UUID` is a `[16]byte`. Instead of calling `MarshalBinary` you should be able to just do:

```go
if uuidVal, ok := v.(uuid.UUID); ok {
    return map[string]any{"uuid": [16]byte(uuidVal)}
}
```

and avoid the `convertToFixedArray` stuff

zeroshade commented:
we're almost there! just a few more comments left. Thanks for your persistence and work on this @badalprasadsingh Once we get this merged I think it'll be good for us to cut a release.

badalprasadsingh committed:
fix: uuid avro type && test for avroLogicalTypes

Signed-off-by: badalprasadsingh <badal@datazip.io>

badalprasadsingh committed:
fix: TestAddFilesFailsSchemaMismatch

Signed-off-by: badalprasadsingh <badal@datazip.io>

badalprasadsingh committed:
Merge main into feat/partitioned-fanout-writer to sync with upstream

badalprasadsingh committed:
avro partition serialization simplified

fix race condition in partitioned fanout writer

Signed-off-by: badalprasadsingh <badal@datazip.io>

badalprasadsingh committed:
merge main

Signed-off-by: badalprasadsingh <badal@datazip.io>

zeroshade commented on manifest.go:
```diff
@@ -1461,33 +1520,116 @@ func mapToAvroColMap[K comparable, V any](m map[K]V) *[]colMap[K, V] {
+}
+
+func convertToFixedArray(bytes []byte, size int) any {
+	arr := reflect.New(reflect.ArrayOf(size, reflect.TypeOf(byte(0)))).Elem()
+	reflect.Copy(arr, reflect.ValueOf(bytes))
```
as with the other sections, I'm still unsure here. This codepath seems to *only* be used during read, so we should never have to produce the `map` representation here. The avro library itself should perform the conversions for us when dealing with fixed size array things.

badalprasadsingh committed:
updated: schema_conversions_test.go

Signed-off-by: badalprasadsingh <badal@datazip.io>

badalprasadsingh committed:
fix: avro deserialization for manifests

Signed-off-by: badalprasadsingh <badal@datazip.io>

badalprasadsingh committed:
fix: manifest.go

Signed-off-by: badalprasadsingh <badal@datazip.io>

badalprasadsingh committed:
fix: minor

Signed-off-by: badalprasadsingh <badal@datazip.io>

badalprasadsingh commented on manifest.go:
```diff
@@ -1546,18 +1689,39 @@ func (d *dataFile) initializeMapData() {
+	if t, ok := v.(time.Time); ok {
+		if logicalType, hasLogical := d.fieldIDToLogicalType[fieldID]; hasLogical {
+			switch logicalType {
+			case avro.Date:
+				days := int32(t.Truncate(24*time.Hour).Unix() / int64((time.Hour * 24).Seconds()))
+
+				return Date(days)
+			case avro.TimestampMicros:
+				return Timestamp(t.UTC().UnixMicro())
```
well, we shouldn't be using `avroPartitionData` while reading, as it's whole purpose is to convert `Iceberg` types -> `Avro Union (map)` format.

badalprasadsingh committed:
fix: minor

Signed-off-by: badalprasadsingh <badal@datazip.io>

badalprasadsingh committed:
fix: deserialization of avro logical types

Signed-off-by: badalprasadsingh <badal@datazip.io>

badalprasadsingh committed:
fix: minor fixes

Signed-off-by: badalprasadsingh <badal@datazip.io>

badalprasadsingh committed:
merge "main" with "feat/partitioned-fanout-writer"

badalprasadsingh commented on manifest_test.go:
```diff
@@ -82,7 +82,7 @@ var (
 				Path:             "/home/iceberg/warehouse/nyc/taxis_partitioned/data/VendorID=null/00000-633-d8a4223e-dc97-45a1-86e1-adaba6e8abd7-00001.parquet",
 				Format:           ParquetFile,
-				PartitionData:    map[string]any{"VendorID": int(1), "tpep_pickup_datetime": time.Unix(1925, 0).UnixMicro()},
+				PartitionData:    map[string]any{"VendorID": int(1), "tpep_pickup_datetime": time.Unix(1925, 0)},
```
Fixed 🙌

badalprasadsingh commented on manifest.go:
```diff
@@ -1546,18 +1689,39 @@ func (d *dataFile) initializeMapData() {
+		}
+		days := int32(t.Truncate(24*time.Hour).Unix() / int64((time.Hour * 24).Seconds()))
+
+		return days
```
Fixed 🙌

badalprasadsingh commented:
Hey @zeroshade, I have updated the workflow.

While writing a manifest, we need to encode it into Avro types. So, just before encoding in the `addEntry()` function, we call `avroPartitionData()`, which converts the data into the respective maps.  

While reading manifests, the decoded Avro types come as Go types, and we use the `convertAvroValueToIcebergType` function inside `initializeMapData()` to convert them back into the respective Iceberg literals.

The `getFieldIDMap()` func is used both during writing manifests and during reading manifests to get the `fieldNameToID`, `fieldIDToType`, `fieldIDToSize` (only during reading) maps. Thus, we need to have `int` type as a case as well (during writing).

badalprasadsingh commented:
For certain types like `decimal` and `uuid`, their primitive Avro types are `fixed`, which is absolutely correct according to the [official Iceberg documentation](https://iceberg.apache.org/spec/#avro). However, according to `hamba/avro`, their preferred primitive Avro types should be `bytes` and `string`, respectively. Therefore, to work with the fixed schema of both types, the use of `maps` becomes a necessity in a union schema.

Specifically, for `decimal` and `uuid` in the `convertAvroValueToIcebergType` function, we get the decoded values as a `map[string]interface{}` for the same reason. For `decimal`, the decoded value is returned as a `big.Rat`. To convert a `big.Rat` into a `DecimalLiteral`, we need to know its scale, which is why we store the decimal scale in the `fieldIDToFixedSize` map before encoding.

zeroshade reviewed (commented):
Just two comments.

I think we're probably okay to merge at this point even if it's not perfect. This is a large enough PR and most of the remaining things I can think of are all performance optimizations and slight refactors that could be done later on while people really want this functionality now. So we can get this out, and then optimize it further afterwards.

So please just address the two comments below and then we can get this merged. 

Once again, thanks so much for staying on top of this. I know it's been a long back-and-forth but this is definitely a great step in the right direction for this project! Thanks!

badalprasadsingh committed:
merge `main` branch with `feat/partitioned-fanout-writer`

badalprasadsingh committed:
Merge `main` branch

zeroshade committed:
add link to issue in hamba

zeroshade committed:
Refine TODO comment for Fixed Schema handling

Updated TODO comment for Fixed Schema in Avro.

zeroshade committed:
fix whitespace issue
